designs/cloudstack-multiple-endpoints.md

```diff
@@ -0,0 +1,177 @@
+# Supporting Cloudstack clusters across endpoints
+
+## Introduction
+
+**Problem:** 
+
+The mangement API endpoint for Cloudstack is a potential points of failure. If one endpoint goes down, then control of all of everything goes down. So, we want to spread our services across many Cloudstack endpoints and hosts to protect against that.
+
+For scalability, multiple Cloudstack endpoints will likely be required for storage and API endpoint throughput for our customer. Just one cluster creation as many as a thousand API calls to ACS (estimated). There are many ways to support this scale, but adding more Cloudstack hosts and endpoints is a fairly foolproof way to do so. Then, there’s the size and performance of the underlying database that each Cloudstack instance runs on.
+
+In CAPC, we are considering addressing the problem by extending our use of the concept of [Failure Domains](https://cluster-api.sigs.k8s.io/developer/providers/v1alpha2-to-v1alpha3.html?highlight=failure%20domain#optional-support-failure-domains) and distributing a cluster across the given ones. However, instead of a failure domain consisting of a zone on a single Cloudstack endpoint, we will redefine it to consist of the unique combination of a Cloudstack zone, api endpoint, account, and domain. In order to support this functionality in EKS-A, we need to have a similar breakdown where an EKS-A cluster can span across multiple endpoints and Failure Domains.
+
+### Tenets
+
+* ****Simple:**** simple to use, simple to understand, simple to maintain
+* ****Declarative:**** intent based system, as close to a Kubernetes native experience as possible
+
+### Goals and Objectives
+
+As a Kubernetes administrator I want to:
+
+* Perform preflight checks when creating/upgrading clusters which span across multiple failure domains
```

Comment:
nit: isn't this an implementation detail?
The user want's to create and upgrade across multiple "zones"
And part of that process involves validating all those zones before continuing

Reply:
I get what you are saying, however I don't think this is a user story. Cluster admins can't "perform validations", at least for now. It's the cli the ones that does it to prevent starting a cluster creation on a misconfigured environment.

I agree that this should be part of the scope of this project. But that's not what this section is about.

If you are firm on including this as part of the user stories, we could maybe rephrase it:

```suggestion
* My cluster and environment across multiple failure domains to be validated before creating/upgrading clusters
```